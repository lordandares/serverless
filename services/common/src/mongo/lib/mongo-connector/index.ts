import mongodb, { MongoClient } from 'mongodb'

let cachedDb: MongoClient

export default async function connect(
  mongoDbUri: string,
): Promise<MongoClient> {
  try {
    if (!mongoDbUri) {
      throw new Error('MongoConnectionError: mongoDbUri is not defined')
    }

    if (cachedDb) {
      return cachedDb
    }

    console.info('mongo:connect')

    const db: MongoClient = await mongodb.MongoClient.connect(mongoDbUri)

    console.info('mongo:connect:success')

    cachedDb = db

    return cachedDb
  } catch (err) {
    throw new Error(`MongoConnectionError: ${err.message}`)
  }
}
