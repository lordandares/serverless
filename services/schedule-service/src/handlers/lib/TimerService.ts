/* istanbul ignore file */

interface Clear {
  id: string
}

interface Set {
  data: object
  id: string
  timestamp: string
}

/* istanbul ignore file */

export function clear({ id }: Clear) {
  // TODO probably fire SNS event or clear cloudwatch timer
  return id
}

export function set({ data, id, timestamp }: Set) {
  // TODO probably fire SNS event or clear cloudwatch timer
  return id
}
