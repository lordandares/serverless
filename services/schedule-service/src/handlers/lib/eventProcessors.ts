import { errors, schemas } from '@lighthouse/serverless-common'
import * as AWS from 'aws-sdk'
import { ClientConfiguration, DocumentClient } from 'aws-sdk/clients/dynamodb'
import SNS from 'aws-sdk/clients/sns'
import * as AWSXRay from 'aws-xray-sdk'
import {
  filter,
  flatMap,
  includes,
  join,
  keys,
  map,
  omit,
  reduce,
} from 'lodash'
import * as moment from 'moment'

declare var process: {
  env: {
    IS_OFFLINE: string
    REGION: string
    OCCURRENCE_RESOLVED_ARN: string
    TABLE_SCHEDULES: string
  }
}

interface ResolvedMatch {
  id: string
  pk: string
  sk: string
  updatedAt: string
}

export const processorStrategies = {
  location: processLocationEvent,
}

export async function processLocationEvent(streamEvent) {
  try {
    if (!streamEvent || !streamEvent.data || !streamEvent.version) {
      throw Error('ProcessLocationEvent: Invalid event received')
    }

    const isOffline = process.env.IS_OFFLINE
    const region = process.env.REGION
    const tableName = process.env.TABLE_SCHEDULES
    const topicArn = process.env.OCCURRENCE_RESOLVED_ARN

    const dynamoDbConfig: ClientConfiguration =
      region === 'localhost'
        ? /* istanbul ignore next */ {
            endpoint: 'http://localhost:8000',
            httpOptions: { timeout: 5000 },
            region,
          }
        : {}

    const DynamoDB = isOffline
      ? AWS.DynamoDB
      : /* istanbul ignore next */
        AWSXRay.captureAWS(AWS).DynamoDB

    const documentClient: DocumentClient = new DynamoDB.DocumentClient(
      dynamoDbConfig,
    )
    const snsClient: SNS = new AWS.SNS()

    const { data, version } = streamEvent

    // data property is a lighthouse event
    const { _id, application, area, user } = data

    console.info(`ProcessLocationEvent: processing ${_id} (${version})`)

    if (version === 'v1') {
      if (!application) {
        throw Error('ProcessLocationEvent: event missing application property')
      }

      const applicationId = application._id
      const patternTypes = ['visit']
      const patternPrefix = `rule-pattern-${applicationId}`

      // NOTE istanbul issue with optional chaining which breaks 100% coverage
      // https://github.com/istanbuljs/istanbuljs/issues/516
      // const areaIds = area && area?.ids || []
      const areaIds = (area && area.ids) || []
      const userId = user

      const patternQueries = flatMap(areaIds, id =>
        map(patternTypes, type => ({
          pk: `${patternPrefix}-${id}`,
          sk: type,
        })),
      )

      console.info(
        `ProcessLocationEvent: created ${
          patternQueries.length
        } pattern queries`,
        {
          patternQueries,
        },
      )

      const resolvedMatches: ResolvedMatch[] = []
      const transactItems: DocumentClient.TransactWriteItem[] = []

      for (const patternQuery of patternQueries) {
        const { pk: rulePatternPk, sk: rulePatternSk } = patternQuery

        const rulePatternQuery: DocumentClient.GetItemInput = {
          Key: { pk: rulePatternPk, sk: rulePatternSk },
          TableName: tableName,
        }

        const rulePatternResponse: DocumentClient.GetItemOutput = await documentClient
          .get(rulePatternQuery)
          .promise()

        const { Item: rulePatternItem } = rulePatternResponse

        if (!rulePatternItem) {
          console.warn('ProcessLocationEvent: unable to find rule pattern', {
            rulePatternQuery,
          })

          continue
        }

        console.info('ProcessLocationEvent: processing rule pattern', {
          rulePatternQuery,
        })

        const {
          matches: rulePatternMatches,
          updatedAt: rulePatternUpdatedAt,
        } = rulePatternItem

        const rulePatternMatchesIds = keys(rulePatternMatches)

        // NOTE: if rule pattern matches is empty then we can safely remove the
        // pattern document. Please note this will only happen on the
        // subsequent rule pattern match even though all matches could be
        // processed initially
        if (rulePatternMatchesIds.length === 0) {
          console.info(
            'ProcessLocationEvent: will remove rule pattern as no matches',
          )

          const rulePatternDelete: DocumentClient.TransactWriteItem = {
            Delete: {
              ConditionExpression: '#updatedAt = :updatedAt',
              ExpressionAttributeNames: {
                '#updatedAt': 'updatedAt',
              },
              ExpressionAttributeValues: {
                ':updatedAt': rulePatternUpdatedAt,
              },
              Key: { pk: rulePatternPk, sk: rulePatternSk },
              TableName: tableName,
            },
          }

          transactItems.push(rulePatternDelete)

          continue
        }

        const processedMatches: string[] = []

        for (const matchId in rulePatternMatches) {
          // NOTE: ensures we don't iterate prototype properties
          /* istanbul ignore next */
          if (!rulePatternMatches.hasOwnProperty(matchId)) {
            continue
          }

          const matchValues = rulePatternMatches[matchId]

          const {
            locationId: matchLocationId,
            pk: matchPk,
            sk: matchSk,
            userId: matchUserId,
          } = matchValues

          const matchesLocationId = !!matchLocationId
            ? includes(areaIds, matchLocationId)
            : true

          const matchesUserId = !!matchUserId ? matchUserId === userId : true

          const matchesEvent = matchesLocationId && matchesUserId

          if (!matchesEvent) {
            console.warn(
              `ProcessLocationEvent: pattern rule does not match data`,
              {
                matchId,
                matchesLocationId,
                matchesUserId,
              },
            )

            continue
          }

          const matchDocumentQuery: DocumentClient.GetItemInput = {
            Key: { pk: matchPk, sk: matchSk },
            TableName: tableName,
          }

          const matchDocumentResponse: DocumentClient.GetItemOutput = await documentClient
            .get(matchDocumentQuery)
            .promise()

          const { Item: matchDocument } = matchDocumentResponse

          if (!matchDocument) {
            console.warn(
              `ProcessLocationEvent: match is missing so will be removed from rule pattern`,
            )

            processedMatches.push(matchId)

            continue
          }

          const {
            endAt: matchEndAt,
            startAt: matchStartAt,
            updatedAt: matchUpdatedAt,
          } = matchDocument

          const now = moment.utc().valueOf()
          const nowIsoString = moment
            .utc()
            .toDate()
            .toISOString()
          const isFutureMatch = now < moment.utc(matchStartAt).valueOf()
          const isExpiredMatch = now > moment.utc(matchEndAt).valueOf()

          if (isFutureMatch) {
            console.warn(
              `ProcessLocationEvent: match can't be processed until ${matchStartAt} so skipping`,
              {
                endIsoString: matchEndAt,
                matchId,
                nowIsoString,
                startIsoString: matchStartAt,
              },
            )

            continue
          }

          if (isExpiredMatch) {
            console.warn(
              `ProcessLocationEvent: match expired at ${matchEndAt} so will be removed from rule pattern`,
              {
                endIsoString: matchEndAt,
                matchId,
                nowIsoString,
                startIsoString: matchStartAt,
              },
            )

            processedMatches.push(matchId)

            continue
          }

          processedMatches.push(matchId)

          resolvedMatches.push({
            id: matchId,
            pk: matchPk,
            sk: matchSk,
            updatedAt: matchUpdatedAt,
          })
        }

        const processedMatchesLength = processedMatches.length

        if (processedMatchesLength) {
          const nextUpdatedAt = moment
            .utc()
            .toDate()
            .toISOString()

          const nextMatches = omit(rulePatternMatches, processedMatches)

          const rulePatternUpdate: DocumentClient.TransactWriteItem = {
            Update: {
              ConditionExpression: '#updatedAt = :updatedAt',
              ExpressionAttributeNames: {
                '#matches': 'matches',
                '#updatedAt': 'updatedAt',
              },
              ExpressionAttributeValues: {
                ':nextMatches': nextMatches,
                ':nextUpdatedAt': nextUpdatedAt,
                ':updatedAt': rulePatternUpdatedAt,
              },
              Key: { pk: rulePatternPk, sk: rulePatternSk },
              TableName: tableName,
              UpdateExpression: `SET #matches = :nextMatches, #updatedAt = :nextUpdatedAt`,
            },
          }

          transactItems.push(rulePatternUpdate)
        }
      }

      const resolvedMatchesLength = resolvedMatches.length
      const transactItemsLength = transactItems.length

      if (transactItemsLength) {
        console.info(
          `ProcessLocationEvent: processing ${transactItemsLength} transactions`,
        )

        const transactParams: DocumentClient.TransactWriteItemsInput = {
          TransactItems: transactItems,
        }

        await documentClient.transactWrite(transactParams).promise()

        console.info(
          'ProcessLocationEvent: completed transactions successfully',
        )
      }

      if (resolvedMatchesLength) {
        console.info(
          `ProcessLocationEvent: processing ${resolvedMatchesLength} resolved matches`,
        )

        for (const resolvedMatch of resolvedMatches) {
          try {
            await snsClient
              .publish({
                Message: JSON.stringify(resolvedMatch),
                TopicArn: topicArn,
              })
              .promise()
          } catch (error) {
            console.error(`ProcessLocationEvent: resolving match failed`, {
              error: error.message,
              resolvedMatch,
            })
          }
        }

        console.info(
          'ProcessLocationEvent: finished processing resolved matches',
        )
      }
    }

    console.info(
      `ProcessLocationEvent: finished processing ${_id} (${version})`,
    )
  } catch (err) {
    throw new errors.ApplicationError({ message: err.message })
  }
}
