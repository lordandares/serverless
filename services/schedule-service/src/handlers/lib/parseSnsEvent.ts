/* istanbul ignore file */

import { SNSEvent } from 'aws-lambda'

import { getOr } from 'lodash/fp'

interface Result {
  body: { [name: string]: any }
  messageId: string
  topicArn: string
}

export default function parseSnsEvent(event: SNSEvent): Result {
  const message = getOr({}, 'Records.0.Sns.Message', event)
  const body = JSON.parse(message)
  const messageId = getOr('unknown', 'Records.0.Sns.MessageId', event)
  const topicArn = getOr('unknown', 'Records.0.Sns.TopicArn', event)

  return { body, messageId, topicArn }
}
