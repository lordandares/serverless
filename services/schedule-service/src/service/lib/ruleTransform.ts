import { schemas } from '@lighthouse/serverless-common'
import * as moment from 'moment'
import { v4 as uuid } from 'uuid'

export interface PayloadToRulePatternDocument {
  actor: schemas.ActorSchema
  existingDocument: schemas.RulePatternDocumentSchema | undefined
  payload: {
    applicationId: string
    locationId: string
    occurrenceId: string
    pk: string
    sk: string
    type: string
  }
}
export function payloadToRulePatternDocument({
  actor,
  existingDocument,
  payload,
}: PayloadToRulePatternDocument): schemas.RulePatternDocumentSchema {
  const { applicationId, locationId, occurrenceId, pk, sk, type } = payload

  const now: string = moment
    .utc()
    .toDate()
    .toISOString()

  const matchValues = { pk, sk }

  const matches: schemas.MatchSchema =
    existingDocument && existingDocument.matches
      ? { ...existingDocument.matches, [occurrenceId]: matchValues }
      : { [occurrenceId]: matchValues }

  const document: schemas.RulePatternDocumentSchema = {
    createdAt: (existingDocument && existingDocument.createdAt) || now,
    createdBy: (existingDocument && existingDocument.createdBy) || actor,
    groupType: 'rule-pattern',
    matches,
    pk: `rule-pattern-${applicationId}-${locationId}`,
    sk: type,
    updatedAt: now,
    updatedBy: actor,
  }

  return document
}
