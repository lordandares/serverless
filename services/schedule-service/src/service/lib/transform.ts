import { schemas } from '@lighthouse/serverless-common'
import * as moment from 'moment'
import { v4 as uuid } from 'uuid'

interface DocumentToSchedulePayloadOptions {
  document: schemas.ScheduleDocumentSchema
}

export function documentToSchedulePayload(
  options: DocumentToSchedulePayloadOptions,
): schemas.ScheduleResponseSchema {
  const { document } = options
  const {
    createdAt,
    createdBy,
    data: {
      areas,
      enabled,
      included,
      locations,
      serviceHours,
      name,
      strategy,
      type,
    },
    endAt,
    scheduleId,
    startAt,
    updatedAt,
    updatedBy,
  } = document

  const payload: schemas.ScheduleResponseSchema = {
    areas,
    createdAt,
    createdBy,
    enabled,
    endAt,
    id: scheduleId,
    included,
    locations,
    name,
    serviceHours,
    startAt,
    strategy,
    type,
    updatedAt,
    updatedBy,
  }

  return payload
}

interface PayloadToScheduleDocumentOptions {
  payload: schemas.SchedulePayloadSchema
  additional: {
    applicationId: string
    createdAt?: string
    createdBy?: schemas.ActorSchema
    id?: string
    included: object
    userId: string
  }
}

export function payloadToScheduleDocument(
  options: PayloadToScheduleDocumentOptions,
): schemas.ScheduleDocumentSchema {
  const { payload, additional } = options
  const {
    areas,
    enabled,
    endAt,
    locations,
    name,
    serviceHours,
    startAt,
    strategy,
    type,
  } = payload

  const {
    applicationId,
    createdAt,
    createdBy,
    id,
    included,
    userId,
  } = additional

  const isNew: boolean = !id

  const now: string = moment
    .utc()
    .toDate()
    .toISOString()

  const actor = { id: userId, label: 'Unknown User', type: 'user' }

  const scheduleId: string = isNew ? uuid() : id

  const document: schemas.ScheduleDocumentSchema = {
    applicationId,
    createdAt: createdAt || now,
    createdBy: createdBy || actor,
    data: {
      areas,
      enabled,
      included,
      locations,
      name,
      serviceHours,
      strategy,
      type,
    },
    endAt,
    groupType: 'schedule',
    pk: `${applicationId}-schedule`,
    scheduleId,
    sk: `${isNew ? now : createdAt}-${scheduleId}`,
    startAt,
    updatedAt: now,
    updatedBy: actor,
  }

  return document
}
