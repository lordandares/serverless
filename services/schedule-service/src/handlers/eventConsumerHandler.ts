import { errors } from '@lighthouse/serverless-common'
import { reduce, sortBy } from 'lodash'

import { processorStrategies } from './lib/eventProcessors'

interface RecordError {
  error: string
  record: string
}

interface StreamEvent {
  data: {
    _id: string
    application: string
    area: string
    timestamp: string
    user: string
  }
  type: string
  version: number
}

export async function eventConsumerHandler(event): Promise<void> {
  try {
    if (!event) {
      throw new errors.ApplicationError({
        message: 'EventsConsumerHandler: missing event',
      })
    }

    const { Records } = event
    const recordErrors: RecordError[] = []
    const totalCount = Records.length

    console.info(`EventsConsumerHandler: Received ${totalCount} records`)

    const streamEvents = reduce(
      Records,
      (accum: StreamEvent[], record) => {
        try {
          const { kinesis } = record
          const json: string = Buffer.from(kinesis.data, 'base64').toString(
            'ascii',
          )
          const parsedJson: StreamEvent = JSON.parse(json)
          accum.push(parsedJson)
        } catch (error) {
          recordErrors.push({
            error: error.message,
            record: JSON.stringify(record),
          })
        }
        return accum
      },
      [],
    )

    // NOTE: ensure stream events ordered by timestamp
    const sortedStreamEvents = sortBy(
      streamEvents,
      streamEvent => streamEvent.data.timestamp,
    )

    let failedCount = 0
    let processedCount = 0

    for (const streamEvent of sortedStreamEvents) {
      // NOTE istanbul issue with optional chaining which breaks 100% coverage
      // https://github.com/istanbuljs/istanbuljs/issues/516
      // const type = streamEvent?.type
      const type = streamEvent && streamEvent.type

      if (!type) {
        continue
      }

      const processorFn = processorStrategies[type]

      if (!processorFn) {
        continue
      }

      try {
        await processorFn(streamEvent)
        processedCount++
      } catch (error) {
        console.error('EventsConsumerHandlerError', { error, streamEvent })
        failedCount++
        continue
      }
    }

    console.info(
      `EventsConsumerHandler: ${processedCount} processed / ${failedCount} failed of ${totalCount} records`,
    )

    if (recordErrors.length) {
      const message = `EventsConsumerHandler: caught ${
        recordErrors.length
      } errors when processing ${Records.length} records`
      console.error(message)

      throw new errors.ApplicationError({
        data: { errors: JSON.stringify(recordErrors) },
        message,
      })
    }
  } catch (err) {
    console.error('EventsConsumerHandlerError', {
      err,
      event,
    })

    throw err
  }
}
