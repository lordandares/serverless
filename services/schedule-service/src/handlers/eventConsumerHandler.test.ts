import { eventConsumerHandler } from './eventConsumerHandler'

import { processorStrategies } from './lib/eventProcessors'

jest.mock('./lib/eventProcessors')

afterEach(() => jest.clearAllMocks())

test('handles no stream event', async () => {
  expect.assertions(1)

  const streamEvent = null

  try {
    await eventConsumerHandler(streamEvent)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[ApplicationError: EventsConsumerHandler: missing event]`,
    )
  }
})

test('handles stream event which fails to parse json but continues to process and throws after', async () => {
  expect.assertions(3)

  const corruptPayload = '{'
  const validPayload = { data: {}, type: 'location' }

  const validEncodedPayload = Buffer.from(
    JSON.stringify(validPayload),
  ).toString('base64')
  const invalidEncodedPayload = Buffer.from(corruptPayload).toString('base64')

  const streamEvent = {
    Records: [
      { kinesis: { data: invalidEncodedPayload } },
      { kinesis: { data: validEncodedPayload } },
    ],
  }
  const errorSpy = jest.spyOn(console, 'error')

  try {
    await eventConsumerHandler(streamEvent)
  } catch (error) {
    expect(processorStrategies.location).toHaveBeenCalledTimes(1)
    expect(errorSpy.mock.calls).toMatchSnapshot()
    expect(error).toMatchSnapshot()
  }
})

test('handles stream event which has no type', async () => {
  expect.assertions(1)

  const payload = { data: {}, type: null }
  const json = JSON.stringify(payload)
  const encodedPayload = Buffer.from(json).toString('base64')
  const streamEvent = { Records: [{ kinesis: { data: encodedPayload } }] }

  await eventConsumerHandler(streamEvent)

  expect(processorStrategies.location).toHaveBeenCalledTimes(0)
})

test('handles stream event which has no type property', async () => {
  expect.assertions(1)

  const payload = { data: {} }
  const json = JSON.stringify(payload)
  const encodedPayload = Buffer.from(json).toString('base64')
  const streamEvent = { Records: [{ kinesis: { data: encodedPayload } }] }

  await eventConsumerHandler(streamEvent)

  expect(processorStrategies.location).toHaveBeenCalledTimes(0)
})

test('handles stream event which has no supported processor function', async () => {
  expect.assertions(1)

  const payload = {
    data: { timestamp: '2020-02-11T03:19:51.508+0000' },
    type: 'testing',
  }
  const json = JSON.stringify(payload)
  const encodedPayload = Buffer.from(json).toString('base64')
  const streamEvent = { Records: [{ kinesis: { data: encodedPayload } }] }

  await eventConsumerHandler(streamEvent)

  expect(processorStrategies.location).toHaveBeenCalledTimes(0)
})

test('handles stream event which has location type', async () => {
  expect.assertions(2)

  const payload = {
    data: { timestamp: '2020-02-11T03:19:51.508+0000' },
    type: 'location',
  }
  const json = JSON.stringify(payload)
  const encodedPayload = Buffer.from(json).toString('base64')
  const streamEvent = { Records: [{ kinesis: { data: encodedPayload } }] }

  await eventConsumerHandler(streamEvent)

  expect(processorStrategies.location).toHaveBeenCalledTimes(1)
  expect(processorStrategies.location).toHaveBeenCalledWith(payload)
})

test('handles stream events which are out of order', async () => {
  expect.assertions(2)

  const payload1 = {
    data: { timestamp: '2020-02-11T03:19:51.508+0000' },
    type: 'location',
  }
  const json1 = JSON.stringify(payload1)
  const encodedPayload1 = Buffer.from(json1).toString('base64')

  const payload2 = {
    data: { timestamp: '2020-02-11T09:19:51.508+0000' },
    type: 'location',
  }
  const json2 = JSON.stringify(payload2)
  const encodedPayload2 = Buffer.from(json2).toString('base64')

  const streamEvent = {
    Records: [
      { kinesis: { data: encodedPayload2 } },
      { kinesis: { data: encodedPayload1 } },
    ],
  }

  await eventConsumerHandler(streamEvent)

  expect(processorStrategies.location).toHaveBeenCalledTimes(2)
  expect(processorStrategies.location).toMatchSnapshot()
})

test('catches event processor errors and continues', async () => {
  expect.assertions(2)

  const payload1 = {
    data: { timestamp: '2020-02-11T03:19:00.000+0000' },
    type: 'location',
  }
  const json1 = JSON.stringify(payload1)
  const encodedPayload1 = Buffer.from(json1).toString('base64')

  const payload2 = {
    data: { timestamp: '2020-02-11T03:20:00.000+0000' },
    type: 'location',
  }
  const json2 = JSON.stringify(payload2)
  const encodedPayload2 = Buffer.from(json2).toString('base64')

  const payload3 = {
    data: { timestamp: '2020-02-11T03:21:00.000+0000' },
    type: 'location',
  }
  const json3 = JSON.stringify(payload3)
  const encodedPayload3 = Buffer.from(json3).toString('base64')

  const streamEvent = {
    Records: [
      { kinesis: { data: encodedPayload1 } },
      { kinesis: { data: encodedPayload2 } },
      { kinesis: { data: encodedPayload3 } },
    ],
  }

  processorStrategies.location
    .mockRejectedValueOnce(new Error('Unknown Error'))
    .mockResolvedValueOnce()
    .mockResolvedValueOnce()

  await eventConsumerHandler(streamEvent)

  expect(processorStrategies.location).toHaveBeenCalledTimes(3)
  expect(processorStrategies.location).toMatchSnapshot()
})

test('catches and rethrows errors', async () => {
  expect.assertions(1)

  try {
    await eventConsumerHandler({})
  } catch (error) {
    expect(error).toEqual(error)
  }
})
