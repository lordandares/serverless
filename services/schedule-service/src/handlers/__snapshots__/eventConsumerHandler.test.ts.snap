// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`catches event processor errors and continues 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": Object {
          "timestamp": "2020-02-11T03:19:00.000+0000",
        },
        "type": "location",
      },
    ],
    Array [
      Object {
        "data": Object {
          "timestamp": "2020-02-11T03:20:00.000+0000",
        },
        "type": "location",
      },
    ],
    Array [
      Object {
        "data": Object {
          "timestamp": "2020-02-11T03:21:00.000+0000",
        },
        "type": "location",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
    Object {
      "isThrow": false,
      "value": Promise {},
    },
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`handles stream event which fails to parse json but continues to process and throws after 1`] = `
Array [
  Array [
    "EventsConsumerHandler: caught 1 errors when processing 2 records",
  ],
  Array [
    "EventsConsumerHandlerError",
    Object {
      "err": [ApplicationError: EventsConsumerHandler: caught 1 errors when processing 2 records],
      "event": Object {
        "Records": Array [
          Object {
            "kinesis": Object {
              "data": "ew==",
            },
          },
          Object {
            "kinesis": Object {
              "data": "eyJkYXRhIjp7fSwidHlwZSI6ImxvY2F0aW9uIn0=",
            },
          },
        ],
      },
    },
  ],
]
`;

exports[`handles stream event which fails to parse json but continues to process and throws after 2`] = `[ApplicationError: EventsConsumerHandler: caught 1 errors when processing 2 records]`;

exports[`handles stream events which are out of order 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": Object {
          "timestamp": "2020-02-11T03:19:51.508+0000",
        },
        "type": "location",
      },
    ],
    Array [
      Object {
        "data": Object {
          "timestamp": "2020-02-11T09:19:51.508+0000",
        },
        "type": "location",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
